connection/connection.go

```diff
@@ -53,170 +53,99 @@ func pipe(c1, c2 PipeConnection) {
 	}
 }
 
-func CreateBasicPlayerConnection(conn Connection, remoteAddr net.Addr) *BasicPlayerConnection {
-	return &BasicPlayerConnection{conn: conn, addr: remoteAddr}
-}
-
-func CreateBasicPlayerConnection2(conn net.Conn, remoteAddr net.Addr) *BasicPlayerConnection {
-	c := CreateBasicConnection2(conn, remoteAddr)
-	return &BasicPlayerConnection{conn: c}
+func CreateBasicPlayerConnection(conn net.Conn, remoteAddr net.Addr) *BasicPlayerConnection {
+	c := CreateBasicConnection(conn)
+	return &BasicPlayerConnection{
+		conn:   c,
+		reader: bufio.NewReader(conn),
+		addr:   remoteAddr,
+	}
 }
 
 // Basic implementation of LoginConnection
 type BasicPlayerConnection struct {
-	conn Connection
-
-	addr    net.Addr
-	hsPk    protocol.Packet
-	loginPk protocol.Packet
-	hs      handshaking.ServerBoundHandshake
-	mcName  string
-
-	hasHS    bool
-	hasHSPk  bool
-	hasLogin bool
-}
+	conn   ByteConnection
+	reader protocol.DecodeReader
 
-func (c *BasicPlayerConnection) ServerAddr() string {
-	hs, _ := c.Handshake()
-	return string(hs.ServerAddress)
+	addr net.Addr
+	hsPk protocol.Packet
+	hs   handshaking.ServerBoundHandshake
 }
 
 func (c *BasicPlayerConnection) ReadPacket() (protocol.Packet, error) {
-	return c.conn.ReadPacket()
+	return protocol.ReadPacket(c.reader)
 }
 
 func (c *BasicPlayerConnection) WritePacket(p protocol.Packet) error {
-	return c.conn.WritePacket(p)
+	pk, _ := p.Marshal() // Need test for err part of this line
+	_, err := c.conn.Write(pk)
+	return err
 }
 
-func (c *BasicPlayerConnection) RemoteAddr() net.Addr {
-	return c.addr
+func (c *BasicPlayerConnection) Handshake() handshaking.ServerBoundHandshake {
+	return c.hs
 }
 
-func (c *BasicPlayerConnection) Handshake() (handshaking.ServerBoundHandshake, error) {
-	if c.hasHS {
-		return c.hs, nil
-	}
-
-	pk, _ := c.HsPk()
-	var err error
-	c.hs, err = handshaking.UnmarshalServerBoundHandshake(pk)
-	if err != nil {
-		return c.hs, err
-	}
-	c.hasHS = true
-	return c.hs, nil
+func (c *BasicPlayerConnection) HsPk() protocol.Packet {
+	return c.hsPk
 }
 
-func (c *BasicPlayerConnection) HsPk() (protocol.Packet, error) {
-	if c.hasHSPk {
-		return c.hsPk, nil
-	}
-	pk, err := c.ReadPacket()
-	if err != nil {
-		return pk, ErrCantGetHSPacket
-	}
+func (c *BasicPlayerConnection) SetHsPk(pk protocol.Packet) {
```

Comment:
SetHandshakePacket(...)

Reply:
I would also change the receiver parameter name to conn instead of just c

Reply:
with all those conns its probably getting harder to know which one is what. if we were to do that, we should probably rename the conn for net.Conn to netConn and the conn for ByteConn to byteConn

Reply:
Sounds like a good idea